src/main/java/io/jenkins/infra/repository_permissions_updater/github_team_sync/YmlTeamLoader.java

```diff
@@ -0,0 +1,64 @@
+package io.jenkins.infra.repository_permissions_updater.github_team_sync;
+import org.yaml.snakeyaml.Yaml;
+import java.io.FileInputStream;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class YmlTeamLoader {
+
+    public static GithubTeamDefinition loadTeam(String filePath) {
+        Path basePath = Paths.get("permissions").toAbsolutePath().normalize();
+        Path resolvedPath = basePath.resolve(filePath).normalize();
+
+        if (!resolvedPath.startsWith(basePath)) {
+            throw new SecurityException("Attempted path traversal out of allowed directory");
+        }
+
+        if (!resolvedPath.toString().endsWith(".yml")) {
+            throw new SecurityException("Invalid file type");
+        }
+
+        if (!Files.exists(resolvedPath)) {
+            throw new RuntimeException("File does not exist: " + resolvedPath);
+        }
+
+        // Load the file
+        try (FileInputStream inputStream = new FileInputStream(resolvedPath.toFile())) {
+
+            Yaml yaml = new Yaml();
+
+            Map<String, Object> data = yaml.load(inputStream);
+            String repoPath = "";
+            String teamName = "";
+            Set<String> developers = new HashSet<>();
+            
+            if (data.containsKey("github")) {
+                repoPath = (String) data.get("github");
+            }
+            
+            if (data.containsKey("github_team")) {
```

Comment:
do we think we need to add this / need it as the only way?

teams are currently done by the repository name and should be easily determinable, (there may be some outliers though and where it might be good to be able to override this)

Reply:
Currently, if a YAML file doesn't include github_team, I don't take any action.
If it does include github_team, I assume there's a unique team name like github_team: abc_team, and update the developers on that team. 

However, based on your description, it seems uncommon to have a unique team name specified. Is it more common for github_team to be included without a specific name, meaning the team name would default to the repository name?

Also, if I have misunderstood anything in the first part, please correct me.

Reply:
it will currently never include a GitHub team you will need to backfill those across the repository with this approach.

(not necessarily a bad thing)

Teams will almost always be unique per repository, occasionally there will be additional teams with access.

e.g. design-library-plugin has the sig-ux team with access.

and there will be other similar examples

Reply:
> e.g. design-library-plugin has the sig-ux team with access.

In particular, while most teams are `${repo}-developers`, there are some "special" teams spanning multiple repositories.

Also, I don't see where team repos are being managed, only team members. Would those special teams be excluded from management as code?

Reply:
I would suggest for non-schema based groups that in the future it is moved to an extra section called: `custom-groups`.

Reply:
Thank you for your guidance. Combining what I've learned from your feedback with TheMeinerLP's suggestions, I'm considering a strategy to improve our current code management. Here's my proposal:

### Repository-Specific Teams (${repo}-developers)
- The members of these teams are specified in the YAML files associated with each repository. 
- To manage these, I plan to backfill the github_team names where necessary and ensure permissions are appropriately set for each team.
  
### Cross-Repository Teams (like sig-ux): 
These teams manage multiple repositories but aren't tied to a specific one, which could lead to consistency issues if managed separately. To address this, I propose:
- **Adding distinct names and permissions** for these special teams directly within our current YAML configuration.

- **Creating a centralized YAML configuration** file exclusively for managing these cross-repository teams. This file would handle all membership and permission settings, ensuring consistency across all repositories they are involved with.

**Below is an example:**

```
github_team:
  name: “abc-developers”
  permissions:
    - write

custom-groups:
  - name: “sig-ux“
    permissions:
      - admin
  - name: “other-team”
    permissions:
      - read
```

Could you please let me know if you think this plan is feasible? I value your experience and guidance highly.

Reply:
yaml file for managing custom teams: sure

the example isn't what I would expect.

I would expect something roughly like:

teams.yml
```yaml
teams:
  - name: sig-ux
     members:
       - timja
       - notmyfault
```

plugin-design-library.yml
```
...
github_teams:
  - name: sig-ux
    permission: admin
```

Reply:
your example is not explained clearly enough

Reply:
Hi @timja ,

Sorry for the confusion in my previous messages. Your comment had not loaded when I replied, which resulted in my responding to an earlier discussion, not directly to your suggestions. 

Here's an updated example:

### permissions/plugin-commons-lang3-api.yml

```
...
developers:
  - "Alaurant"
  - "TheMeinerLP"
github_team: 
  name: "plugin-commons-lang3-api-developers"
  permission: "Read"
custom_groups:
  - name: “sig-ux“
    permissions: "admin"
  - name: “other-team”
    permissions: "read"
```
Or, if there are no custom groups:

```
...
developers:
  - "Alaurant"
  - "TheMeinerLP"
github_team: 
  name: "plugin-commons-lang3-api-developers"
  permission: "Read"
custom_groups:
```
**Note:** The **“custom_groups”** can range from zero to any number. However, the **“github_team”** section must not be empty if developers are listed. This includes specifying the **team name**, whether it corresponds to a **team name** in the YAML file or a **team** in the **${repo}-developers** format within the organization. Each **“github_team”** is associated with the **developers** specified in the YAML file. 

In the permission/*.yaml files, we manage general teams (github_team), which include team names, permissions, and developers. For special teams, we manage only permissions that are related to specific repositories, specifying which teams are associated with this repo and what their permissions are. The management of developers for special teams is handled in the github_teams/*.yaml files.

```
 - RPU
   ...
   - folder: permissions
   - folder: github-teams （new）
   - folder: teams
   ...
```
### github-teams/sig-ux.yml
```
name: sig-ux
members:
   - timja
   - notmyfault
```

### github-teams/sig-app.yml
```
name: sig-app
members:
  - "Alaurant"
  - "TheMeinerLP"
```

Considering this structure and the content in the new **"github-teams”** folder closely mirrors the current **”teams”** folder, is there a possibility we could leverage the existing structure instead of creating a new one? 

I look forward to your reply and guidance on this matter.

Reply:
I think custom_groups could just be dropped and github_team made a list which would simplify it

Reply:
Thanks for your suggestions on the YAML structure. I want to make sure I understand your proposal correctly and implement it as you envision. To clarify, are you suggesting:

## A. We list permissions for each repository directly in a permissions.yaml file, specifying the teams related to each repo?

```
...
developers:
  - "Alaurant"
  - "TheMeinerLP"
github_teams: 
  - name: "plugin-commons-lang3-api-developers"
    permission: "Read"
  - name: “sig-ux“
    permissions: "admin"
  - name: “security”
    permissions: "read"
```

## B. Create a new single YAML file that contains all special team definitions and keep general teams in permissions/*.yaml.

### teams.yml

```
teams:
  - name: sig-ux
     members:
       - "timja"
       - "notmyfault"
   - name: sig-app
     members:
       - "Alaurant"
       - "TheMeinerLP"
   ...
```
       
### permissions/plugin-design-library.yml

```
...
github_team:
  name: sig-ux
  permission: admin
```

## C. We set up a new folder, say github-teams, with a separate YAML file for each team.


> ```
>  - RPU
>    ...
>    - folder: permissions
>    - folder: github-teams （new）
>    - folder: teams
>    ...
> 
> ```

> ### github-teams/sig-ux.yml
> 
> ```
> name: sig-ux
> members:
>    - timja
>    - notmyfault
>    
> ```
> ### github-teams/company-cloudbees-developers.yml
> 
> ```
> name: company-cloudbees-developers
> members:
>    - jtnord
>    - ...
> ```

Could you let me know which one of these options aligns with your vision, or if there's another approach you were considering?

Reply:
please read through the solutions you've put a bit closer, none of them look quite workable but are getting closer

Reply:
Thank you for your feedback. I appreciate you taking the time to review the solutions.

However, I'm not entirely sure where my current solution might be lacking, but I've made some adjustments based on my understanding. I've highlighted the changes in the **New**:

### A. We list permissions for each repository directly in a permissions.yaml file, specifying the teams related to each repo?

### permission/plugin-commons-lang3-api.yml
```
              ...
              developers:
                - "Alaurant"
                - "TheMeinerLP"
              github_teams: 
                - name: "plugin-commons-lang3-api-developers"
                   permission: "read"
**New**            default: "yes" # Indicates this is a general team, should sync with the developers listed above
                - name: “sig-ux“
                   permissions: "admin"
**New**            default: "no" # Indicates this is a special team
                - name: “security”
                   permissions: "read"
**New**            default: "no"
```

### B. Create a new single YAML file that contains all special team definitions and keep general teams in permissions/*.yml.
### teams.yml
```
              teams:
                - name: sig-ux
                   members:
                     - "timja"
                     - "notmyfault"
**New**     roles: 
**New**        - repo: "plugin-design-library"
**New**          permission: "admin"
**New**       - repo: "plugin-commons-lang3-api"
**New**          permission: "read"
                 - name: sig-app
                    members:
                      - "Alaurant"
                      - "TheMeinerLP"
**New**      roles:
**New**        - repo: "plugin-design-library"
**New**          permission: "admin"
**New**       - repo: "plugin-commons-lang3-api"
**New**          permission: "read"
   ...
```
### permissions/plugin-design-library.yml
```
...
github_team:
  name: "plugin-commons-lang3-api-developers"
  permission: "Read"
```

### C. We set up a new folder, say github-teams, with a separate YAML file for each team.
```
 - RPU
   ...
   - folder: permissions
   - folder: github-teams （new）
   - folder: teams
   ...

```
### github-teams/sig-ux.yml
```
                     name: sig-ux
                     members:
                       - timja
                       - notmyfault
**New**      roles:
**New**        - repo: "plugin-design-library"
**New**          permission: "admin"
**New**       - repo: "plugin-commons-lang3-api"
**New**          permission: "read"
   
```
### github-teams/company-cloudbees-developers.yml

```
                     name: company-cloudbees-developers
                     members:
                       - jtnord
                       - ...
**New**      roles:
**New**        - repo: "company-cloudbees"
**New**           permission: "write"
```


If these changes are still not workable, could you please provide some hints on which areas might need more consideration?

Reply:
I don't think that teams should be defined in a place like C and then in that place reference which repositories it has permissions on.

I should be able to look at a repository file and see who has permission on it.

I would define the 'custom' teams as in C.

Option A looks mostly workable.

Another option would be something like:

```
repository_team: `design-library-plugin-developers`
additional_github_teams:
  - name: sig-ux
     role: admin
```

Reply:
Thank you for your feedback. I will try your solution. I also have a couple of questions to clarify:

Q1: In your example, **repository_team** doesn't specify a role. Does this mean all  **repository_team** will have the same role? If so, what would that role be? read, triage, write, maintain, or admin?

Q2: For **additional_github_teams**, I noticed that we aren't managing the **developers** here. Where should the developers be managed, or is it assumed that they won't be managed through code?

Thanks again.

Reply:
1. repository_team should always be admin.

2. those should be managed in one place, e.g. in a `github_teams` folder, they can't be managed in a per repository file as that doesn't make sense.

Reply:
Thank you for the guidance.

Regarding the github_teams folder, I noticed that RPU currently has a folder structure similar to what we might need, located under teams. Are these teams the same as the special teams we discussed? I am considering enhancing this by adding a **”github_team_name“** tag for synchronization. Here are the current setups I've found:

<img width="1212" alt="image" src="https://github.com/user-attachments/assets/31f43e59-432d-47af-8a61-8551fd62dbb2">

<img width="597" alt="image" src="https://github.com/user-attachments/assets/0ce31b4f-f97b-489b-8323-dc235a3b7a01">

I'm thinking of structuring it like this:

```
---
name: "ux"
github_team_name: "sig-ux" （new）
developers:
  - "timja"
  - "notmyfault"
```

**Alternatively**, would it be better to create a new github_teams folder that includes both the team name and developers?

### github_teams/sig-ux.yml

```
name: sig-ux
developers:
  - "timja"
  - "notmyfault"
```

Looking forward to your thoughts.

Reply:
I don't think there's a need for a github_team_name, just rename the ux team to be sig-ux and just have the names match

Reply:
I’m providing a summary of our recent extensive discussion to ensure that we mutually understand and agree on our conclusions. Below, I detail the changes to the YAML file structures and potential methods for their implementation:

### permission/*.yml

```
...
developers:
  - "Alaurant"
  - "TheMeinerLP"
repository_team: design-library-plugin-developers  # New entry, needs to be backfilled
additional_github_teams:  # New entry, requires backfilling; can include any number of teams
  - name: sig-ux
    role: admin
...
```
### teams/*.yml

```
---
name: "ux"
github_team_name: "sig-ux"  # New entry; may not be backfilled. The name should be added manually.
developers:
  - "timja"
  - "notmyfault"
```

### Summary in Words:
In the **permissions/*.yml** file, the **repository_team** field is linked to a list of **“developers”**, who are the members of the repository team, and this team is automatically assigned an **admin** role. Additionally, the **additional_github_teams** are recorded by their name and role within this repository.
The developers for these additional_github_teams are managed separately in the **teams/*.yml files**. And, each team is represented by a YAML file.

Please let me know if there is anything that differs from your understanding~

Reply:
right one critical issue here is that a lot of peoples LDAP usernames do not match with their GitHub username.

e.g. 
for James Nord his GitHub is `@jtnord`'s but his ldap is `teilo`
https://github.com/jenkins-infra/repository-permissions-updater/blob/afb732d06d78bc6af7e44cd359b52e2ee0e6effb/teams/core.yml#L16

Reply:
Thank you for the feedback. I will attempt to address this issue using the structured format you suggested:

```
developers:
  - ldap: timja
     github: timja
  - ldap: teilo
     github: jtnord
```
Additionally, I've noticed that the number of developers listed in the YAML file doesn't match the number in the GitHub team. Like these:

![question_devs](https://github.com/user-attachments/assets/794686cf-c0fd-481d-98b0-20c42fafc034)
![question 2](https://github.com/user-attachments/assets/739eea2c-8899-4db5-a1a7-64faa34c4feb)

Since noticing the difference, it got me thinking:

Could some developers be listed only in the GitHub team or only in LDAP?

If every developer must have an account on both sides, and observing the difference in developers' numbers, I wonder which source should be considered authoritative when their counts differ—the YAML or GitHub? If neither, could you provide some guidelines on how to handle this?

Reply:
The yaml should be authoritative but there will need to be a one-off backfill from GitHub to the code

Reply:
Thank you for your clear response.

Since this is a one-off backfill, does that mean it should be completed separately before this function is put into use? If that's the case, I'll go ahead and create a new PR. In it, I plan to outline some common issues, such as mismatches in numbers and names, and I’ll propose potential solutions for us to discuss further.

Additionally, considering the frequency of our discussions in PRs might be triggering numerous notifications, should we consider using a different platform or tool to avoid overwhelming you? The decision is yours.

Thank you again for guiding me through this process.

Reply:
It needs to be done before go-live of the new solution, ideally as close to when the other solution is ready to merge in case changes occur again.

Reply:
I've set up a new repository for the one-off backfill and updated the README with the current backfill logic and a few questions. When you have a moment, I'd appreciate it if you could take a look: https://github.com/Alaurant/OneOffSyncTool.

Reply:
> Extended Permissions

1. You should be able to see all GitHub permissions with this report which might help with your permissions request: https://reports.jenkins.io/github-jenkinsci-permissions-report.json

> Format

2. I wouldn't use the empty string if you can't match a person: `github: ""`, just omit the key.

> Clarification of Special Teams:

3. I'm not sure what you mean by clarification of special teams, yes its name is `SIG: UX`

> Challenges with Multiple YAML Files for a Single Repository

4. I think only one file should be managing the GitHub permissions ideally good point

Reply:
Thank you for your patience in reading and responding. The permissions report will be helpful.

> Format

Following your advice to omit the key, would the entries look like this?
```
- ldap: Alice
- github: Bob
- ldap: Chalie
  github: Chalie
```

> Clarification of Special Teams:

I appreciate your help on the "SIG: UX" clarification. Could you confirm if the following matches are correct?

1. **cloudbees-developers**: `company-cloudbees-developers`
2. **core**:  `Core` or `Core PR Reviewers`**?**
3. **security**: `core-security-review`
4. **openshift-developers**: Which team corresponds to this in the provided image?

<img width="990" alt="image" src="https://github.com/user-attachments/assets/b452cb6c-8469-4c6c-82d8-b15eea5d8afd">

Thank you again.

Reply:
1. Yes entries would look like that
2. yes cloudbees is company-cloudbees...
3. core is Core
4. I don't think security is core-security-review, they are separate
5. there's no openshift-developers team in GitHub but I'm sure it would be a good idea.